### PR TITLE
QN Update election round on successful council election

### LINKED ISSUE #3573: QN inconsistency in the elections
There has been some inconsistency on the elections data returned by the QN on the olympia testnet:

### I. Rogue vote detected by user that they did not do (Joystream/pioneer#2753)
5HT8Mty5U5hFKVwv24uFswX2e41vurU3F5rkLMoETR8XmjFy

```gql
query Wrong_Votes_Bug {
  voteCastEvents(where: {inBlock_lte: 43200, AND: {castVote: {castBy_eq: "5HT8Mty5U5hFKVwv24uFswX2e41vurU3F5rkLMoETR8XmjFy"}}}) {
    castVote {
      stake
      voteFor {
        memberId
        stake
        status
      }
      castBy
    }
    inBlock
  }
}
```

### II. The `councilElections` field is empty on the elected council
the currently elected council has an empty councilElections field and on the previous council nextCouncilElections.isFinished is false.

```gql
query {
  electedCouncils {
    isResigned
    electedAtTime
    councilElections { isFinished }
    nextCouncilElections { isFinished }
  }
}
```

Screenshot from 2022-04-07 12-37-19

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201709380879215/1202093277058287) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by mnaamani**
Fixes https://github.com/Joystream/joystream/issues/3573 and https://github.com/Joystream/joystream/issues/3727

For testing I ran a specific scenario with testing runtime profile:

```ts
import electCouncil from '../flows/council/elect'
import { scenario } from '../Scenario'

scenario('Council', async ({ job }) => {
  const firstCouncilJob = job('electing council', electCouncil)
  const secondCouncilJob = job('electing second council', electCouncil).requires(firstCouncilJob)
}
```

At the end of the scenario (immediately) ran query:

```graphql
query {
  electionRounds (orderBy: [createdAt_DESC] limit: 3) {
    cycleId
    isFinished
    electedCouncil { id isResigned }
    nextElectedCouncil{ id }
  }
  electedCouncils (orderBy: [createdAt_DESC] limit: 3) {
    id
    councilElections { cycleId }
  }
}
```

and got expected results:
```json
{
  "data": {
    "electionRounds": [
      {
        "cycleId": 13,
        "isFinished": true, // <--
        "electedCouncil": {
          "id": "00000001", // prior elected council, that was active during this election cycle 13
          "isResigned": true
        },
        "nextElectedCouncil": {
          "id": "00000002" // the second council that was elected in this cycle 13
        }
      },
      {
        "cycleId": 12,
        "isFinished": true,
        "electedCouncil": {
          "id": "00000000",
          "isResigned": true
        },
        "nextElectedCouncil": {
          "id": "00000001" // council elected in cycle 12
        }
      },
      {
        "cycleId": 11,
        "isFinished": true,
        "electedCouncil": {
          "id": "00000000",
          "isResigned": true
        },
        "nextElectedCouncil": null
      }
    ],
    "electedCouncils": [
      {
        "id": "00000002",
        "councilElections": [
          {
            "cycleId": 13
          }
        ]
      },
      {
        "id": "00000001",
        "councilElections": [
          {
            "cycleId": 12
          }
        ]
      },
      {
        "id": "00000000",
        "councilElections": []
      }
    ]
  }
}
```

### CODE REVIEW

**Review by thesan — APPROVED** [review]
LGTM !

I keep on getting confused with `ElectionRound.ElectedCouncil` being the council elected in the previous election :face_with_spiral_eyes:

File: query-node/mappings/src/council.ts [hunk 1]
```diff
@@ -479,13 +497,16 @@ export async function council_NewCouncilElected({ event, store }: EventContext &
     electedAtBlock: event.blockNumber,
     electedAtTime: new Date(event.blockTimestamp),
     electedAtNetwork: CURRENT_NETWORK,
-    councilElections: oldElectedCouncil?.nextCouncilElections || [],
-    nextCouncilElections: [],
+    councilElections: oldElectedCouncil?.nextCouncilElections || [], // always empty ?
+    nextCouncilElections: [], // always empty ?
```
Comment (Lezek123) [inline query-node/mappings/src/council.ts]: Those fields are derived, so I think it's best to leave them empty.
Also `oldElectedCouncil?.nextCouncilElections` should always be empty at this point, so it really makes no difference.
```suggestion
    councilElections: [],
    nextCouncilElections: [],
```

File: query-node/mappings/src/council.ts [hunk 2]
```diff
@@ -324,8 +342,8 @@ export async function council_AnnouncingPeriodStarted({ event, store }: EventCon
   // specific event processing
 
   // restart elections
-  const electedCouncil = await getCurrentElectedCouncil(store)
-  await startNextElectionRound(store, electedCouncil, event)
+  const { cycleId } = await concludeElectionRound(store, event)
```
Comment (Lezek123) [inline query-node/mappings/src/council.ts]: The election is already concluded when the idle period begins, here it is redundantly "concluded again" at the end of the idle period.

**Review by Lezek123 — CHANGES_REQUESTED** [review]
I also suggest adding new assertions in the integration tests that would verify the query node state of the election rounds etc.

One more thing I noticed is the undeterministic values in `updatedAt` for `ElectedCouncil`, `ElectionRound` etc., which seem not to be based on block timestamp:

```
    "electedCouncils": [
      {
        "id": "00000000",
        "updatedAt": "2022-06-03T09:58:16.458Z",
        "endedAtTime": "2022-06-03T09:58:12.000Z",
        // ...
        "nextCouncilElections": [
          {
            "updatedAt": "2022-06-03T09:56:18.541Z",
            "endedAtTime": "2022-06-03T09:55:52.000Z",
            // ...
```

In theory Hydra should auto-fill those values with a current block timestamp, but I guess there could be some reasons the database entities get updated in a way that's not caught and acted-upon by the framework.

File: query-node/mappings/src/council.ts (hunk 1 above)
Comment (mnaamani) [reply query-node/mappings/src/council.ts]: Thanks for clarifying this. Yes I did notice they were derived in the schema definitions, and I did ask why we set them at all if they are indeed derived https://github.com/Joystream/joystream/issues/3573#issuecomment-1143211675

File: query-node/mappings/src/council.ts (hunk 2 above)
Comment (mnaamani) [reply query-node/mappings/src/council.ts]: Fixed in dee72ad30edd0e7c7839d58ce89303549c324816

**Comment by mnaamani** [comment]
> I also suggest adding new assertions in the integration tests that would verify the query node state of the election rounds etc.

That is a good idea, but I have to admit, this is my Achilles heel. I can't make sense of how to write any query code in ts and in our test framework. Can I ask you to maybe help with this.

**Comment by mnaamani** [comment]
> One more thing I noticed is the undeterministic values in `updatedAt` for `ElectedCouncil`, `ElectionRound` etc., which seem not to be based on block timestamp:
> 
> ```
>     "electedCouncils": [
>       {
>         "id": "00000000",
>         "updatedAt": "2022-06-03T09:58:16.458Z",
>         "endedAtTime": "2022-06-03T09:58:12.000Z",
>         // ...
>         "nextCouncilElections": [
>           {
>             "updatedAt": "2022-06-03T09:56:18.541Z",
>             "endedAtTime": "2022-06-03T09:55:52.000Z",
>             // ...
> ```
> 
> In theory Hydra should auto-fill those values with a current block timestamp, but I guess there could be some reasons the database entities get updated in a way that's not caught and acted-upon by the framework.

I think perhaps you forgot that council id `0` is the bootstrapped council so the timestamp is when the query-node processor was strated? https://github.com/Joystream/joystream/blob/d9d9b49bb376039f4021a53ca903fc075acfb8c6/query-node/mappings/src/bootstrap.ts#L68

**Pull request merged**